LibC+AK: Various improvements to wchar_t, mbrtowc, and related functions

timschumi opened the pull request:
Some minor improvements to `mbrtowc`, added tests for `mbrtowc` and `mbsinit`, and the necessary `wchar_t` formatter in AK.

Hendiadyoin1 commented on AK/Format.h:
```diff
@@ -423,6 +423,13 @@ struct Formatter<char> : StandardFormatter {
+struct Formatter<wchar_t> : StandardFormatter {
+    void format(FormatBuilder& builder, wchar_t value)
+    {
+        builder.put_u64(value, 16, true);
```
considering `wchar`s are characters maybe at least try to do an acscii- or utf-lookup

timschumi replied:
Do you have any examples on how I would do an UTF-8 lookup (ASCII I can probably manage myself)? Also, I don't see this being used much apart from tests, and for test things I'd actually prefer printing it as hexadecimal.

Hendiadyoin1 replied:
I am sadly not very knowledgeable with whchars, I only know that "they can address a bigger character set", but i do not know which "bigger" char-set they address here.
If you know that maybe make an Array, where you lookup the wchar and get the utf-8 code points for it

timschumi replied:
The way that `wchar_t` is implemented right now is that one `wchar_t` contains one code point (i.e. the up to 21 bytes that are stored in a single UTF-8 sequence). So I'd basically just need a way of telling the system "here is a code point, now print it as a unicode character".

IdanHo replied:
StringBuilder has `append_code_point()` if that's helpful.

timschumi replied:
Yeah, that seems to work! I'll push a new revision using that.

timschumi commented:
Updated the `wchar_t` formatter to use `append_code_point()` and to fall back to using the `unsigned long` formatter for hexadecimal, decimal, and other formats. I also made some minor adjustments to the `mbsinit` test case to properly handle the return type.

bgianfo commented:
Can you add a test for the AK `whar_t` formatter change? 
See: `Tests/AK/TestFormat.cpp`

timschumi commented:
Updated the AK commit to include tests. I also changed the formatter to do the codepoint conversion in a seperate StringBuilder so that I can use the String formatter for the advanced format options.

kleinesfilmroellchen commented on AK/Format.cpp:
```diff
@@ -688,6 +688,19 @@ void Formatter<char>::format(FormatBuilder& builder, char value)
+void Formatter<wchar_t>::format(FormatBuilder& builder, wchar_t value)
+{
+    if (m_mode == Mode::Binary || m_mode == Mode::BinaryUppercase || m_mode == Mode::Decimal || m_mode == Mode::Octal || m_mode == Mode::Hexadecimal || m_mode == Mode::HexadecimalUppercase) {
+        Formatter<unsigned long> formatter { *this };
```
This and the following line would read nicer with u64.

timschumi replied:
I now realized that I somehow mixed up the size of `wchar_t` with the `ul` that was required for some of the return values in the tests. I chose `unsigned long` because I wanted a type that would be 4 bytes on x86 and 8 bytes on x86_64, but after testing it seems like `wchar_t` is always 4 bytes in size. I'll amend the commit appropriately to use `u32`.

kleinesfilmroellchen reviewed (commented):
I think the formatter looks fine. Somebody with more POSIX knowledge should look at the wchar stuff again (I'm not super knowledgeable).

timschumi commented on Tests/LibC/TestWchar.cpp:
```diff
@@ -43,3 +43,31 @@ TEST_CASE(wcscoll)
+    size_t ret = mbrtowc(nullptr, "\xdf", 1, &state);
+
+    if (ret != -2ul)
+        FAIL(String::formatted("mbrtowc accepted partial multibyte sequence with return code {} (expected -2)", static_cast<ssize_t>(ret)));
```
`(size_t) -1` is another possible return value (hooray for POSIX), so I wanted to display the actual return value as a signed value instead of unsigned (since it's more unlikely that we will ever consume enough bytes to reach that value the other way around, at least for this test case).

timschumi commented:
Rebased onto current master, no other changes.
